sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceStrategy.scala

```diff
@@ -643,6 +647,34 @@ object DataSourceStrategy {
     (nonconvertiblePredicates ++ unhandledPredicates, pushedFilters, handledFilters)
   }
 
+  def translateAggregate(aggregates: AggregateExpression): Option[AggregateFunc] = {
+
+    def columnAsString(e: Expression): String = e match {
+      case AttributeReference(name, _, _, _) => name
+      case Cast(child, _, _) => child match {
```

Comment:
Is this safe? Aggregating casted values and casting aggregated values looks different.

Reply:
I guess this is probably OK?
For example, if I have `sum(cast(SALARY as bigInt))` here, I will remove cast and push down `sum(SALARY)` to data source. Then I will cast the output of `sum(SALARY)` to bigInt.

Reply:
What if `SALARY` is a tiny int? Looks `sum(SALARY)` can have an overflow easily in the datasource side.

Reply:
For example, if `t` exists in a database side, it seems we can get different answers with/without pushdown?
```
scala> Seq(0.6, 0.7, 0.8).toDF("v").write.saveAsTable("t")
scala> sql("select sum(cast(v as int)) from t").show()
+-------------------+
|sum(CAST(v AS INT))|
+-------------------+
|                  0|
+-------------------+
```

Reply:
Thanks for the example. I actually only want to strip off the cast added by Spark. For example, when doing sum, Spark cast integral type to long. 
```
case Sum(e @ IntegralType()) if e.dataType != LongType => Sum(Cast(e, LongType))
```
For the casting added by Spark, I will remove the casting, push down aggregate and do the same casting on database side.
If the cast is from user, I will keep the cast and NOT push down `aggregate (cast(col))` for now.
To differentiate user explicitly casting from Spark added casting, I will add a flag some where. Does this sound OK to you?

Reply:
For example, how about the fractional case? If `e` is float, we should not remove the cast but pass it into a database side for computing double-typed partial aggregated values?
```
case Sum(e @ FractionalType()) if e.dataType != DoubleType => Sum(Cast(e, DoubleType))
```

Reply:
If the col data type is float, I think we should remove the cast and pass it to database side. On database side, cast it to double:
```
if (cast by Spark)
  remove cast from sum
  push down Sum(col) to database
  if (datatype of col is float)
    cast to Double, something like SELECT  Sum(CAST(col AS DOUBLE)) FROM table
else // cast by user
  not remove cast and not push down aggregate
```

Reply:
@maropu @huaxingao 

such casting are removed by spark in https://github.com/apache/spark/pull/31079